src/traces/table/calc.js

```diff
@@ -0,0 +1,131 @@
+    var headerHeight = headerRowHeights.reduce(function(a, b) {return a + b;}, 0);
+    var scrollHeight = groupHeight - headerHeight;
+    var minimumFillHeight = scrollHeight + c.uplift;
+    var anchorToRowBlock = makeAnchorToRowBlock(rowHeights, minimumFillHeight);
+    var anchorToHeaderRowBlock = makeAnchorToRowBlock(headerRowHeights, headerHeight);
+    var headerRowBlocks = makeRowBlock(anchorToHeaderRowBlock, []);
+    var rowBlocks = makeRowBlock(anchorToRowBlock, headerRowBlocks);
+    var uniqueKeys = {};
+    var columnOrder = trace._fullInput.columnorder;
+    var columnWidths = trace.header.values.map(function(d, i) {
+        return Array.isArray(trace.columnwidth) ?
+            trace.columnwidth[Math.min(i, trace.columnwidth.length - 1)] :
+            isFinite(trace.columnwidth) && trace.columnwidth !== null ? trace.columnwidth : 1;
+    });
+    var totalColumnWidths = columnWidths.reduce(function(p, n) {return p + n;}, 0);
+
+    // fit columns in the available vertical space as there's no vertical scrolling now
+    columnWidths = columnWidths.map(function(d) {return d / totalColumnWidths * groupWidth;});
+
+    var calcdata = {
+        key: trace.index,
+        translateX: domain.x[0] * gd._fullLayout._size.w,
+        translateY: gd._fullLayout._size.h - domain.y[1] * gd._fullLayout._size.h,
+        size: gd._fullLayout._size,
+        width: groupWidth,
+        height: groupHeight,
+        columnOrder: columnOrder, // will be mutated on column move, todo use in callback
+        groupHeight: groupHeight,
+        rowBlocks: rowBlocks,
+        headerRowBlocks: headerRowBlocks,
+        scrollY: 0, // will be mutated on scroll
+        cells: trace.cells,
+        headerCells: trace.header,
+        gdColumns: trace.header.values.map(function(d) {return d[0];}),
+        gdColumnsOriginalOrder: trace.header.values.map(function(d) {return d[0];}),
+        prevPages: [0, 0],
+        scrollbarState: {scrollbarScrollInProgress: false},
+        columns: trace.header.values.map(function(label, i) {
+            var foundKey = uniqueKeys[label];
+            uniqueKeys[label] = (foundKey || 0) + 1;
+            var key = label + '__' + uniqueKeys[label];
+            return {
+                key: key,
+                label: label,
+                specIndex: i,
+                xIndex: columnOrder[i],
+                xScale: xScale,
+                x: undefined, // initialized below
+                calcdata: undefined, // initialized below
+                columnWidth: columnWidths[i]
+            };
+        })
+    };
+
+    calcdata.columns.forEach(function(col) {
+        col.calcdata = calcdata;
+        col.x = xScale(col);
+    });
+
+    return wrap(calcdata);
+};
+
+function xScale(d) {
+    return d.calcdata.columns.reduce(function(prev, next) {
+        return next.xIndex < d.xIndex ? prev + next.columnWidth : prev;
+    }, 0);
+}
+
+function makeRowBlock(anchorToRowBlock, auxiliary) {
```

Comment:
Might be nice to reuse some of the `heatmap/` calc routines. For example `clean2dArray` for cells.values.

@monfera have you looked into `Heatmap.calc`?

Reply:
Thanks, I'll look into reusability there!